tests/kernel/mem_protect/userspace/src/main.c

```diff
@@ -0,0 +1,235 @@
+}
+
+static void read_kernram(void)
+{
+	/* Try to read from kernel RAM. */
+	void *p = _current->init_data;
+
+	printk("%p\n", p);
+	zassert_unreachable("Read from kernel RAM did not fault\n");
+}
+
+static void write_kernram(void)
+{
+	/* Try to write to kernel RAM. */
+	_current->init_data = NULL;
+	zassert_unreachable("Write to kernel RAM did not fault\n");
+}
+
+extern int _k_neg_eagain;
+
+static void write_kernro(void)
+{
+	/* Try to write to kernel RO. */
+	_k_neg_eagain = -EINVAL;
+	zassert_unreachable("Write to kernel RO did not fault\n");
+}
+
+static void write_kerntext(void)
+{
+	/* Try to write to kernel text. */
+	memcpy(&_is_thread_essential, 0, 4);
+	zassert_unreachable("Write to kernel text did not fault\n");
+}
+
+__kernel static int kernel_data;
+
+static void read_kernel_data(void)
+{
+	/* Try to read from embedded kernel data. */
+	int value = kernel_data;
+
+	printk("%d\n", value);
+	zassert_unreachable("Read from __kernel data did not fault\n");
+}
+
+static void write_kernel_data(void)
+{
+	kernel_data = 1;
+	zassert_unreachable("Write to  __kernel data did not fault\n");
+}
+
+/*
+ * volatile to avoid compiler mischief.
+ */
+volatile int *ptr = NULL;
+#if defined(CONFIG_X86)
+volatile size_t size = MMU_PAGE_SIZE;
+#elif defined(CONFIG_ARM) && defined(CONFIG_PRIVILEGED_STACK_SIZE)
+volatile size_t size = CONFIG_PRIVILEGED_STACK_SIZE;
+#else
+volatile size_t size = 512;
+#endif
+
+static void read_kernel_stack(void)
+{
+	/* Try to read from kernel stack. */
+	int s[1];
+
+	s[0] = 0;
+	ptr = &s[0];
+	ptr = (int *)((unsigned char *)ptr - size);
+	printk("%d\n", *ptr);
+	zassert_unreachable("Read from kernel stack did not fault\n");
+}
+
+static void write_kernel_stack(void)
+{
+	/* Try to write to kernel stack. */
+	int s[1];
+
+	s[0] = 0;
+	ptr = &s[0];
+	ptr = (int *)((unsigned char *)ptr - size);
+	*ptr = 42;
+	zassert_unreachable("Write to kernel stack did not fault\n");
+}
+
+
+static struct k_sem sem;
+
+static void pass_user_object(void)
+{
+	/* Try to pass a user object to a system call. */
+	k_sem_init(&sem, 0, 1);
+	zassert_unreachable("Pass a user object to a syscall did not fault\n");
+}
+
+__kernel static struct k_sem ksem;
+
+static void pass_noperms_object(void)
+{
+	/* Try to pass a object to a system call w/o permissions. */
+	k_sem_init(&ksem, 0, 1);
+	zassert_unreachable("Pass an unauthorized object to a "
+			    "syscall did not fault\n");
+}
+
+__kernel struct k_thread kthread_thread;
+
+#define STACKSIZE 512
+K_THREAD_STACK_DEFINE(kthread_stack, STACKSIZE);
+
+void thread_body(void)
+{
+}
+
+static void start_kernel_thread(void)
+{
+	/* Try to start a kernel thread from a usermode thread */
+	k_thread_create(&kthread_thread, kthread_stack, STACKSIZE,
+			(k_thread_entry_t)thread_body, NULL, NULL, NULL,
+			K_PRIO_PREEMPT(1), K_INHERIT_PERMS,
+			K_NO_WAIT);
+	zassert_unreachable("Create a kernel thread did not fault\n");
+}
+
+void test_main(void *unused1, void *unused2, void *unused3)
```

Comment:
This is wrong here (and in protection/src/main.c, from which it was copied).
Should just be test_main(void).
Not sure how that happened; either I messed up when I converted these tests over to ztest in the first place or the test_main() prototype changed along the way.

Reply:
Fix for protection test in https://github.com/zephyrproject-rtos/zephyr/pull/5175